lucene/core/src/java/org/apache/lucene/search/QueryVisitor.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.util.Set;
+import java.util.function.Supplier;
+
+import org.apache.lucene.index.Term;
+import org.apache.lucene.util.automaton.Automaton;
+
+/**
+ * Allows recursion through a query tree
+ *
+ * @see Query#visit(QueryVisitor)
+ */
+public abstract class QueryVisitor {
+
+  /**
+   * Called by leaf queries that match on a specific term
+   *
+   * @param term  the term the query will match on
+   */
+  public void matchesTerm(Term term) { }
+
+  /**
+   * Called by leaf queries that match terms against an Automaton
+   * @param query               the leaf query
+   * @param field               the field to match against
+   * @param isBinary            {@code true} if the Automaton is built to match against byte[] rather than char[]
+   * @param automatonSupplier   a supplier for the built Automaton
+   */
+  public void matchesAutomaton(Query query, String field, boolean isBinary, Supplier<Automaton> automatonSupplier) {}
+
+  /**
+   * Called by leaf queries that do not match on terms
+   * @param query the query
+   */
+  public void visitLeaf(Query query) { }
+
+  /**
+   * Pulls a visitor instance for visiting child clauses of a query
+   *
+   * Returns {@code null} if the child clauses should not be visited
+   *
+   * The default implementation returns {@code this}, unless {@code occur} is equal
+   * to {@link BooleanClause.Occur#MUST_NOT} in which case it returns
+   * {@code null}
+   *
+   * @param occur   the relationship between the parent and its children
+   * @param parent  the query visited
+   */
+  public QueryVisitor getSubVisitor(BooleanClause.Occur occur, Query parent) {
```

Comment:
not sure I like the null here. if we keep this logic I think it should return an `Optional<QueryVisitor>` or always return non-null

Reply:
I'm not sure I like `Optional`...  We use `null` elsewhere in the codebase to signal "stop doing stuff here", for example in `Weight.scorer()`, so it seems an established pattern to me?

Reply:
ok

Reply:
I hate "Optional" unless we get value types in Java.